### PR TITLE
Correctly handle denied access in the firewall

### PR DESCRIPTION
**Pull request by aschempp**
This now fixes and improves a lot of things 🙈 

1. Fixes the authentication system to correctly render the Contao login page even if the current route is not a Contao page (e.g. a custom controller)
1. Fixes the `ContaoLoginAuthenticator` from rendering the wrong page for login (if the user is not fully authenticated).
2. ~Fixes the remember me authentication system~ (see https://github.com/contao/contao/pull/6815)
3. Adds an `AccessDeniedHandler` to render the 403 page via the firewall exception handler
4. Correctly enforces fully authentication for changing personal data, changing the user password and configuring two-factor authentication
5. Removes the ExceptionConverter and PrettyErrorScreen from rendering 401 and 403 pages, because these are rendered by the firewall

There are two "behaviour changes" I can think of
-  obviously the fully-authentication works and is now enforces as described in point 4
- If no 401 or 403 pages are configured in root page, we now render the generic error (through the pretty error screen listener) instead of the generic Symfony 401/403 message. Since we are in Contao frontend scope, I think that should be fine.

### TODO:
- [x] ~Needs a migration for existing RememberMe tokens (the migration tool cannot migrate binary data to string value because the binary data is null-padded and therefore longer than the new string field).~

### CODE REVIEW

File: core-bundle/contao/modules/ModuleLogin.php [hunk 1]
```diff
@@ -161,7 +161,12 @@ protected function compile()
 			$this->Template->hasError = true;
 			$this->Template->message = $GLOBALS['TL_LANG']['ERR']['invalidTwoFactor'];
 		}
-		elseif ($exception instanceof AuthenticationException)
+		elseif ($isRememberMe && $exception instanceof InsufficientAuthenticationException)
+		{
+			$this->Template->hasError = true;
+			$this->Template->message = $GLOBALS['TL_LANG']['ERR']['insufficientAuthentication'];
+		}
+		elseif ($exception instanceof AuthenticationException && !$exception instanceof InsufficientAuthenticationException)
```
Comment (Toflar) [inline core-bundle/contao/modules/ModuleLogin.php]: And what if `!$isRememberMe && $exception instanceof InsufficientAuthenticationException`?

Comment (aschempp) [reply core-bundle/contao/modules/ModuleLogin.php]: Then we do not want to show a message. The `InsufficientAuthenticationException` is added by our own authenticator, only for the case to detect that re-login was requested by the security authenticator.

File: core-bundle/src/Entity/RememberMe.php [hunk 2]
```diff
@@ -32,10 +32,10 @@ class RememberMe
     #[GeneratedValue]
     protected int $id;
 
-    #[Column(type: 'binary_string', length: 88, nullable: false, options: ['fixed' => true])]
+    #[Column(length: 88, nullable: false)]
     protected string $series;
 
-    #[Column(type: 'binary_string', length: 64, nullable: false, options: ['fixed' => true])]
+    #[Column(length: 64, nullable: false)]
```
Comment (leofeyer) [inline core-bundle/src/Entity/RememberMe.php]:
```suggestion
    #[Column(length: 88, nullable: false)]
```

File: manager-bundle/skeleton/config/security.yaml [hunk 3]
```diff
@@ -33,6 +33,7 @@ security:
             request_matcher: contao.routing.frontend_matcher
             provider: contao.security.frontend_user_provider
             user_checker: contao.security.user_checker
+            access_denied_handler: contao.security.access_denied_handler
```
Comment (Toflar) [inline manager-bundle/skeleton/config/security.yaml:36]: You have to also update the `README.md` in the `core-bundle` for regular installations.

**Review by Toflar — CHANGES_REQUESTED** [review]

File: manager-bundle/skeleton/config/security.yaml (hunk 3 above)
Comment (leofeyer) [reply manager-bundle/skeleton/config/security.yaml:36]: And the `core-bundle/tests/Functional/app/config/security.yaml` file.

Comment (aschempp) [reply manager-bundle/skeleton/config/security.yaml:36]: fixed in 9bb19bdf3b732c6a6938d11cacdabf033bc1929f

**Review by Toflar — DISMISSED** [review]
Sure needs some manual testing which I currently don't have the time for but from a logical standpoint, the code looks good to me.

File: core-bundle/src/Entity/RememberMe.php (hunk 2 above)
Comment (aschempp) [reply core-bundle/src/Entity/RememberMe.php]: see https://github.com/contao/contao/pull/6815

**Review by Toflar — DISMISSED** [review]

File: core-bundle/contao/templates/modules/mod_login.html5 [hunk 4]
```diff
@@ -15,6 +15,9 @@
       <input type="hidden" name="REQUEST_TOKEN" value="<?= $this->requestToken ?>">
       <input type="hidden" name="_target_path" value="<?= $this->targetPath ?>">
       <input type="hidden" name="_always_use_target_path" value="<?= $this->forceTargetPath ?>">
+      <?php if ($this->rememberMe): ?>
+        <input type="hidden" name="autologin" value="1">
```
Comment (leofeyer) [inline core-bundle/contao/templates/modules/mod_login.html5:19]: What does this do?

File: core-bundle/contao/languages/en/default.xlf [hunk 5]
```diff
@@ -128,6 +128,9 @@
       <trans-unit id="ERR.invalidLogin">
         <source>Login failed (note that usernames and passwords are case-sensitive)!</source>
       </trans-unit>
+      <trans-unit id="ERR.insufficientAuthentication">
+        <source>Full authentication is required to access this resource.</source>
```
Comment (leofeyer) [inline core-bundle/contao/languages/en/default.xlf]: The message is pretty technical and normal users might not know what "full authentication" is or what "this resource" is referring to. Can we think of a more helpful way of saying it?

**Review by leofeyer — COMMENTED** [review]
The code works fine, however, on the 401 page, I can log in as a different user now. Shouldn‘t we remove the `username` field in this case?

File: core-bundle/contao/languages/en/default.xlf (hunk 5 above)
Comment (aschempp) [reply core-bundle/contao/languages/en/default.xlf]: Yeah I could not think of anything so I took the original exception message. But it felt ok, don't we get similar messages when reauthentication is required e.g. on GitHub?

Comment (aschempp) [reply core-bundle/contao/languages/en/default.xlf]: Message is now `Please re-authenticate to access this resource.`

**Comment by aschempp** [comment]
You are right. Username field removed in 9dc8d9e9fe242eee735ebfaaf4a9555d44cb53fc

I also manually tested the two-factor authentication, which was bumpy even for the regular login 🙈

File: core-bundle/contao/languages/en/default.xlf [hunk 6]
```diff
@@ -128,6 +128,9 @@
       <trans-unit id="ERR.invalidLogin">
         <source>Login failed (note that usernames and passwords are case-sensitive)!</source>
       </trans-unit>
+      <trans-unit id="ERR.insufficientAuthentication">
+        <source>Please re-authenticate to access this resource.</source>
```
Comment (leofeyer) [inline core-bundle/contao/languages/en/default.xlf]:
```suggestion
        <source>Please enter your password to continue.</source>
```

WDYT?

File: core-bundle/contao/templates/modules/mod_login.html5 [hunk 7]
```diff
@@ -15,6 +15,10 @@
       <input type="hidden" name="REQUEST_TOKEN" value="<?= $this->requestToken ?>">
       <input type="hidden" name="_target_path" value="<?= $this->targetPath ?>">
       <input type="hidden" name="_always_use_target_path" value="<?= $this->forceTargetPath ?>">
+      <?php if ($this->remembered): ?>
+        <input type="hidden" name="username" value="<?= $this->value ?>">
```
Comment (leofeyer) [inline core-bundle/contao/templates/modules/mod_login.html5:19]: This looks like something that can be manipulated. Since there is a user already, can we not omit this entirely?

Comment (aschempp) [reply core-bundle/contao/templates/modules/mod_login.html5:19]: we cannot, because the username must be submitted for the firewall authenticator to accept it. However, I don't think manipulation is a problem here, because if you enter a different username, you would need to know the password of that user, and you would simply be logged in as someone else then.

Comment (leofeyer) [reply core-bundle/contao/templates/modules/mod_login.html5:19]: > must be submitted for the firewall authenticator to accept it

Since we have our own `ContaoLoginAuthenticator`, we could get the username from the token in the `getCredentials()` method if it has not been submitted, right?

But yeah, I‘m unsure if this is necessary. The remembered user could also submit the regular login form with another username. 🤔

File: core-bundle/contao/modules/ModuleLogin.php [hunk 8]
```diff
@@ -105,20 +112,14 @@ protected function compile()
...
-		{
-			$authUtils = $container->get('security.authentication_utils');
-			$exception = $authUtils->getLastAuthenticationError();
-			$lastUsername = $authUtils->getLastUsername();
-		}
-
-		$authorizationChecker = $container->get('security.authorization_checker');
-
-		if ($authorizationChecker->isGranted('ROLE_MEMBER'))
+		if ($request && $user instanceof FrontendUser && !$isTwoFactorInProgress && (!$isRemembered || $objPage->type != 'error_401'))
```
Comment (leofeyer) [inline core-bundle/contao/modules/ModuleLogin.php]: Why did you add the `$objPage->type != 'error_401'` check here?

Comment (leofeyer) [reply core-bundle/contao/modules/ModuleLogin.php]: Also, the `$request` check seems unrelated.

The code translates to: Show the "you are logged in as …" message if

* the user is a `FrontendUser`
* and we are not in the middle of a 2FA process
* and the user is fully authenticated **or the page is not a 401 page**.

This means that the same login form behaves differently on a regular page and on a 401 page. If an installation uses the 401 page to redirect to a regular page with a login form, the reauthentication options are never shown.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.
